parsl/monitoring/monitoring.py

```diff
@@ -280,17 +280,22 @@ def monitor_wrapper(f,
         Wrap the Parsl app with a function that will call the monitor function and point it at the correct pid when the task begins.
         """
         def wrapped(*args, **kwargs):
+            command_q = Queue(maxsize=10)
             p = Process(target=monitor,
                         args=(os.getpid(),
                               task_id,
                               monitoring_hub_url,
                               run_id,
+                              command_q,
                               logging_level,
                               sleep_dur),
                         name="Monitor-Wrapper-{}".format(task_id))
             p.start()
             try:
-                return f(*args, **kwargs)
+                res = f(*args, **kwargs)
+                command_q.put("Finished")
```

Comment:
If `f` throws an exception, then this added sequence does not happen, and the `finally p.terminate()` happens.

That means that a function which fails very fast (which is a common failure mode) will still not be reported.

Reply:
I'm also a little concerned about `put` blocking if the monitoring process no longer exists (for example, if ctrl-C has been pressed and killed some of the processes). I haven't investigated this in depth, but https://docs.python.org/3/library/multiprocessing.html talks about some potential deadlocks.

Reply:
> If `f` throws an exception, then this added sequence does not happen, and the `finally p.terminate()` happens.
> 
> That means that a function which fails very fast (which is a common failure mode) will still not be reported.

I tested the situation that `f` fails by adding `return 1 / 0` before `command_q.put("Finished")`. The added sequence did not happen like what you expected. 

So my new change is to add an `except` block that also does the sequence and raises the exception. 

I did not put the sequence in the `finally`  block, because I think the `p.terminate()` is useful to ensure the monitoring process will exit anyway.

Reply:
I think it is fine to put it in the finally block. Maybe even better to.

The code as it is now will try the "except put/join" if f completes successfully but the first put/join sequence raises an exception.

Putting the put/join code in `finally`, with only the invocation of `f` in the try body, will behave differently: it will catch exceptions from invoking `f`, but will not catch exceptions from invoking put/join.

I think it is better to keep the try block as tightly focused around the invocation of `f` as possible.

To keep the p.terminate() in there even if put/join throws an exception, wrap the whole thing in a new `try/finally` block.

Reply:
Fixed now